bibliopixel/drivers/spi_driver_base.py

```diff
@@ -1,75 +1,158 @@
-                a = int(s[0])
-                b = int(s[1])
+    def send_packet(self, data):
+        raise NotImplementedError
 
-            if a < 0 or b < 0:
-                error(BAD_FORMAT_ERROR)
+    def compute_packet(self, data):
+        return data
 
-            self._bootstrapSPIDev()
-            import spidev
-            self.spi = spidev.SpiDev()
-            self.spi.open(a, b)
-            self.spi.max_speed_hz = int(self._spiSpeed * 1000000.0)
-            log.info('py-spidev speed @ %.1f MHz',
-                     (float(self.spi.max_speed_hz) / 1000000.0))
-        else:
-            self._buf = bytearray(self._buf)
-            self.spi = self._open(self.dev, 'wb')
-
-    def _bootstrapSPIDev(self):
-        if self.use_py_spi:
-            try:
-                import spidev
-            except:
-                error(CANT_IMPORT_SPIDEV_ERROR)
-
-        if not os.path.exists(self.dev):
+
+# partial source @see https://armbedded.taskit.de/node/318
+# @see <linux/spi/spidev.h>
+# hex numbers from good old strace(1)
+SPI_IOC_WR_MODE = 0x40016b01
+SPI_IOC_RD_MODE = 0x80016b01
+SPI_IOC_WR_BITS_PER_WORD = 0x40016b03
+SPI_IOC_RD_BITS_PER_WORD = 0x80016b03
+SPI_IOC_WR_MAX_SPEED_HZ = 0x40046b04
+SPI_IOC_RD_MAX_SPEED_HZ = 0x80046b04
+
+
+class SpiFileInterface(SpiBaseInterface):
+    """ using os open/write to send data"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        if not os.path.exists(self._dev):
             error(CANT_FIND_ERROR)
 
+        self._fnctl = fcntl
+        self._spi = open(self._dev, "wb")
+
+        speed = self._set_speed(self._spi_speed)
+
+        bits = self._set_bits_per_word(8)
+        log.info('file io spi speed @ %.1f MHz, %d bits per word', speed / 1e6, bits)
+
+    def _set_speed(self, speed):
+        speed_b = struct.pack(">I", int(speed * 1e9))  # unint32
+        self._fnctl.ioctl(self._spi, SPI_IOC_WR_MAX_SPEED_HZ, speed_b)
+
+        self._fnctl.ioctl(self._spi, SPI_IOC_RD_MAX_SPEED_HZ, speed_b)
+        speed = struct.unpack(">I", speed_b)[0] / 1000
+        return speed
+
+    def _set_bits_per_word(self, bits):
+        bits_b = struct.pack("B", bits)  # uint8
+        self._fnctl.ioctl(self._spi, SPI_IOC_WR_BITS_PER_WORD, bits_b)
+
+        self._fnctl.ioctl(self._spi, SPI_IOC_RD_MAX_SPEED_HZ, bits_b)
+        bits = int.from_bytes(bits_b, byteorder='big')
+        return bits
+
+    def send_packet(self, data):
+        self._set_speed(self._spi_speed)
+        package_size = 4032  # bit smaller than 4096 because of headers
+        for i in range(int(math.ceil(len(data) / package_size))):
+            start = i * package_size
+            end = (i + 1) * package_size
+            self._spi.write(data[start:end])
+            self._spi.flush()
+
+
+class SpiPyDevInterface(SpiBaseInterface):
+    """ using py-spidev to send data"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        a, b = -1, -1
```

Comment:
The next dozen lines have some aspects of lameness to them.  :-P

`a` and `b` are mysteries until the end.  You catch it when `self._dev` is `1.2.3` and throw the right exception, but not when it's `x.y`.

[This commit](https://github.com/rec/BiblioPixel/commit/0e8562a002a20399eaa32795a3f55b8aad566192) is shorter and handles all possible failures.